op-program/host/host.go

```diff
@@ -51,83 +58,193 @@ func FaultProofProgram(logger log.Logger, cfg *config.Config) error {
 		kv = kvstore.NewDiskKV(cfg.DataDir)
 	}
 
-	var preimageOracle preimage.OracleFn
-	var hinter preimage.HinterFn
-	if cfg.FetchingEnabled() {
-		logger.Info("Connecting to L1 node", "l1", cfg.L1URL)
-		l1RPC, err := client.NewRPC(ctx, logger, cfg.L1URL)
-		if err != nil {
-			return fmt.Errorf("failed to setup L1 RPC: %w", err)
+	var getPreimage func(key common.Hash) ([]byte, error)
+	var hinter func(hint string) error
+	if IsHost() {
+		if cfg.FetchingEnabled() {
+			logger.Info("Connecting to L1 node", "l1", cfg.L1URL)
+			l1RPC, err := client.NewRPC(ctx, logger, cfg.L1URL)
+			if err != nil {
+				return fmt.Errorf("failed to setup L1 RPC: %w", err)
+			}
+
+			logger.Info("Connecting to L2 node", "l2", cfg.L2URL)
+			l2RPC, err := client.NewRPC(ctx, logger, cfg.L2URL)
+			if err != nil {
+				return fmt.Errorf("failed to setup L2 RPC: %w", err)
+			}
+
+			l1ClCfg := sources.L1ClientDefaultConfig(cfg.Rollup, cfg.L1TrustRPC, cfg.L1RPCKind)
+			l2ClCfg := sources.L2ClientDefaultConfig(cfg.Rollup, true)
+			l1Cl, err := sources.NewL1Client(l1RPC, logger, nil, l1ClCfg)
+			if err != nil {
+				return fmt.Errorf("failed to create L1 client: %w", err)
+			}
+			l2Cl, err := sources.NewL2Client(l2RPC, logger, nil, l2ClCfg)
+			if err != nil {
+				return fmt.Errorf("failed to create L2 client: %w", err)
+			}
+			l2DebugCl := &L2Source{L2Client: l2Cl, DebugClient: sources.NewDebugClient(l2RPC.CallContext)}
+
+			logger.Info("Setting up pre-fetcher")
+			prefetch := prefetcher.NewPrefetcher(l1Cl, l2DebugCl, kv)
+			getPreimage = func(key common.Hash) ([]byte, error) { return prefetch.GetPreimage(ctx, key) }
+			hinter = prefetch.Hint
+		} else {
+			logger.Info("Using offline mode. All required pre-images must be pre-populated.")
+			getPreimage = kv.Get
+			hinter = func(hint string) error {
+				logger.Debug("ignoring prefetch hint", "hint", hint)
+				return nil
+			}
 		}
+	}
 
-		logger.Info("Connecting to L2 node", "l2", cfg.L2URL)
-		l2RPC, err := client.NewRPC(ctx, logger, cfg.L2URL)
+	// Setup client I/O for preimage oracle interaction
+	var pClientRW *readWritePair
+	if IsHost() {
+		var pHostRW *readWritePair
+		var err error
+		pClientRW, pHostRW, err = bidirectionalPipe()
 		if err != nil {
-			return fmt.Errorf("failed to setup L2 RPC: %w", err)
+			return fmt.Errorf("failed to create preimage pipe: %w", err)
 		}
+		oracleServer := preimage.NewOracleServer(pHostRW)
+		launchOracleServer(logger, oracleServer, getPreimage)
+		defer pHostRW.Close()
+	} else {
+		r := os.NewFile(pClientRFd, "pClientR")
+		w := os.NewFile(pClientWFd, "pClientW")
+		pClientRW = &readWritePair{r, w}
```

Comment:
Can these not be combined into the same file?

Reply:
It's nicer this way. Allows the `readWritePair` to be abstracted as a `io.ReadWriter` so we just pass a single object to the oracle/prefetcher.

Reply:
this is no longer relevant. bidirectional pipes for host/client communication are now being used